src/metabase/models/setting.clj

```diff
@@ -1485,14 +1486,14 @@
   [visibilities]
   ;; ignore Database-local values, but not User-local values
   (binding [*database-local-values* nil]
-    (into
-     {}
-     (comp (filter (fn [[_setting-name setting]]
-                     (and (not (database-local-only? setting))
-                          (can-read-setting? setting visibilities))))
-           (map (fn [[setting-name]]
-                  [setting-name (get setting-name)])))
-     @registered-settings)))
+    (let [setting-names (into []
+                              (keep (fn [[setting-name setting]]
+                                      (when (and (not (database-local-only? setting))
+                                                 (can-read-setting? setting visibilities))
+                                        setting-name)))
+                              @registered-settings)
+          values (perf/pmap get setting-names)]
```

Comment:
I'm a bit nervous about parallelism here - we have a significant number of settings, and this is a very high traffic API, could create a lot of contention.

I wonder whether a cache wouldn't be more appropriate, we can invalidate it whenever settings are updated, and likewise our settings poll the database in the background and they could also bust the cache when they pick something up. That would give a much more significant speed-up without so much thread coordination.

Reply:
You are right to be paranoid. Even though the FJP (which used under beneath parallelStream) is designed for many short-lived tasks, it makes sense to implement the caching that you suggested to worry less about possible contention.

However, how difficult would it be to make such a cache? I see that this settings system is quite an intricate one, and the data is spread across many places. Do you have something specific in mind where such cache can be placed (and how it can react to values changing)?

Reply:
Unfortunately my suggestion was made blindly without specific ideas, haven't looked at the relevant code in a while.